### PR TITLE
Valid anagram (Resolves #94)

### LINKED ISSUE #94: Valid Anagrams Checker
# Valid Anagrams Checker

## Problem Description
Write a function that determines if two strings are anagrams of each other. Two strings are considered anagrams if they contain exactly the same characters with the same frequency, regardless of their order.

## Function Specification
Determines if two strings are anagrams of each other.

## Examples
###Example 1:

Input: str1 = "anagram", str2 = "nagaram"
Output: true

###Example 2:

Input: str1 = "rat", str2 = "car"
Output: false

## Resources
- [Leetcode Solution](https://leetcode.com/problems/valid-anagram/solutions/)

### PR DESCRIPTION
**Pull request by fatima-malik99**
---
name: solution review 
about: A template PR for code review with a checklist
---

A function that determines if two strings are anagrams of each other. Two strings are considered anagrams if they contain exactly the same characters with the same frequency, regardless of their order.

# Behavior
## Input:
- Two strings (`str1` and `str2`), where each string can contain:
  - Alphabetic characters (both uppercase and lowercase)
  - Numbers
  - Special characters
  - Spaces
  - Empty strings

## Output:
- A boolean value (`True` or `False`) where:
  - `True`: if the strings are valid anagrams of each other
  - `False`: if the strings are not anagrams

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by MadiMalik** [comment]
Overall, great job, Fatima! Your code and documentation are clean, well-structured, and very understandable. I just wanted to point out that your function is case-sensitive, meaning W is not equal to w. However, you don’t have a test case specifically addressing this behavior. Additionally, I think adding a boundary test to handle long inputs would be beneficial. Once again, excellent work!

File: solutions/tests/test_valid_anagram.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+Created on: 2025-01-11
+@author: Fatima
+"""
+
+import unittest
+from solutions.valid_anagram import valid_anagram
+
+
+class Testvalidanagram(unittest.TestCase):
```
Comment (mudassra-taskeen) [inline solutions/tests/test_valid_anagram.py]: The class name is not in PascalCase. Rename the class to `TestValidAnagram` to follow PascalCase convention.

**Review by mudassra-taskeen — CHANGES_REQUESTED** [review]

File: solutions/tests/test_valid_anagram.py (hunk 1 above)
Comment (fatima-malik99) [reply solutions/tests/test_valid_anagram.py]: Ohh, thankyou so much for notifying me. I didn't notice it.

File: solutions/tests/test_valid_anagram.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    def test_not_anagram(self):
+        """It should return False for non-anagram strings."""
+        self.assertFalse(valid_anagram("hello", "world"))
+
+    def test_same_word(self):
+        """It should identify same words as anagrams."""
+        self.assertTrue(valid_anagram("word", "word"))
+
+    def test_empty_string(self):
+        """It should identify empty strings as amagrams."""
```
Comment (mudassra-taskeen) [inline solutions/tests/test_valid_anagram.py]: There is a docstring typo in test_empty_string. It should be anagrams instead of amagrams

File: solutions/tests/test_valid_anagram.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        self.assertTrue(valid_anagram("", ""))
+
+    def test_case_sensitive(self):
```
Comment (mudassra-taskeen) [inline solutions/tests/test_valid_anagram.py:41]: I like the fact that you added defensive cases like invalid inputs, but you should also include performance tests for extremely large strings.

File: solutions/valid_anagram.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+    # Create character frequency dictionary
+    char_count = {}
+
+    # Count occurrences of each character in str1
+    for char in str1:
+        char_count[char] = char_count.get(char, 0) + 1
+
+    # Decrease character counts and check if any character is missing or overused
+    for char in str2:
```
Comment (mudassra-taskeen) [inline solutions/valid_anagram.py]: Fatima, Use `first_argument `and `second_argument` in the `with self.assertRaises` tests for  more clarity.

**Review by mudassra-taskeen — CHANGES_REQUESTED** [review]

File: solutions/tests/test_valid_anagram.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+    def test_basic_anagram(self):
+        """It should identify basic anagrams correctly."""
+        self.assertTrue(valid_anagram("listen", "silent"))
+
+    def test_not_anagram(self):
+        """It should return False for non-anagram strings."""
+        self.assertFalse(valid_anagram("hello", "world"))
+
+    def test_same_word(self):
```
Comment (mudassra-taskeen) [inline solutions/tests/test_valid_anagram.py:37]: I noticed that test_same_word is redundant since the test_basic_anagram already covers identical strings. You should remove test _same_word to avoid the redundancy.

**Review by mudassra-taskeen — CHANGES_REQUESTED** [review]

File: solutions/tests/test_valid_anagram.py (hunk 1 above)
Comment (fatima-malik99) [reply solutions/tests/test_valid_anagram.py]: I've just updated the name please have a look.

File: solutions/tests/test_valid_anagram.py (hunk 2 above)
Comment (fatima-malik99) [reply solutions/tests/test_valid_anagram.py]: Thank you for notifying me. I've corrected it now.

File: solutions/tests/test_valid_anagram.py (hunk 3 above)
Comment (fatima-malik99) [reply solutions/tests/test_valid_anagram.py:41]: Hey Mudassara, thanks for the suggestion! I’m keeping the focus on functional and edge case testing for now. Since performance concerns haven't come up, I’d prefer to keep performance tests separate if needed later. Let me know if you see specific inefficiencies!

File: solutions/tests/test_valid_anagram.py (hunk 5 above)
Comment (fatima-malik99) [reply solutions/tests/test_valid_anagram.py:37]: Hey Mudassara, I see your point! I’m keeping test_same_word because it ensures that identical words are correctly handled as anagrams without unnecessary processing. But I appreciate your input!

**Comment by fatima-malik99** [comment]
> Overall, great job, Fatima! Your code and documentation are clean, well-structured, and very understandable. I just wanted to point out that your function is case-sensitive, meaning W is not equal to w. However, you don’t have a test case specifically addressing this behavior. Additionally, I think adding a boundary test to handle long inputs would be beneficial. Once again, excellent work!

Hi Madiha thanks for your review. I already have a test_case_sensitive for handling it. Also, I will definitely add boundary case for long inputs. Thank you for your feedback.

File: solutions/valid_anagram.py (hunk 4 above)
Comment (fatima-malik99) [reply solutions/valid_anagram.py]: Thank you for the suggestion Mudassara. But I prefer keeping the test cases concise. Since the values are already simple and self-explanatory, I’d rather avoid adding extra lines. Let me know if you have any other feedback!

Comment (mudassra-taskeen) [reply solutions/valid_anagram.py]: These variable names are good, but the issue is that they are not describing function properly. Moreover, they should be in Snakecase according to checklist. 
You can replace str1 with first argument and str 2 with second argument.

Comment (fatima-malik99) [reply solutions/valid_anagram.py]: Ohh sorry for misunderstanding, I got your point now let me update it.

Comment (fatima-malik99) [reply solutions/valid_anagram.py]: I've updated the variable names, please check now.

File: solutions/tests/test_valid_anagram.py (hunk 1 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_valid_anagram.py]: Yeah, now it fulfills the PascalCase convention

File: solutions/valid_anagram.py (hunk 4 above)
Comment (mudassra-taskeen) [reply solutions/valid_anagram.py]: Great, now those variable names are up to mark. They are describing function clearly.

File: solutions/tests/test_valid_anagram.py (hunk 5 above)
Comment (mudassra-taskeen) [reply solutions/tests/test_valid_anagram.py:37]: Okay, your point of view is also valid.

**Comment by mudassra-taskeen** [comment]
> > Overall, great job, Fatima! Your code and documentation are clean, well-structured, and very understandable. I just wanted to point out that your function is case-sensitive, meaning W is not equal to w. However, you don’t have a test case specifically addressing this behavior. Additionally, I think adding a boundary test to handle long inputs would be beneficial. Once again, excellent work!
> 
> Hi Madiha thanks for your review. I already have a test_case_sensitive for handling it. Also, I will definitely add boundary case for long inputs. Thank you for your feedback.


I have not seen any boundary case in your code. You should add them for better performance.

**Comment by fatima-malik99** [comment]
> > > Overall, great job, Fatima! Your code and documentation are clean, well-structured, and very understandable. I just wanted to point out that your function is case-sensitive, meaning W is not equal to w. However, you don’t have a test case specifically addressing this behavior. Additionally, I think adding a boundary test to handle long inputs would be beneficial. Once again, excellent work!
> > 
> > 
> > Hi Madiha thanks for your review. I already have a test_case_sensitive for handling it. Also, I will definitely add boundary case for long inputs. Thank you for your feedback.
> 
> I have not seen any boundary case in your code. You should add them for better performance.

Please check again I added boundary case after Madiha's suggestion.

**Review by mudassra-taskeen — APPROVED** [review]
I have checked both files. Your solution has no error, working fine, and is ready to merge.
